Client provides visitors with a collection of articles

1CIM opened the pull request:
## PT: https://www.pivotaltracker.com/story/show/178091891

### User Story
```
As a visitor
In order to get my daily dose of conspiracies
I need the site to show me the most recent conspiracy articles
```

### Changes suggested
- add `axios` & semantic
- Send a GET request to the articles endpoint `/api/articles`
- General layout of the page
- A hero section with the first entry of articles
- Article Cards with the rest of entries
- Display error if 500 error

### What we learned
- Use modules
- Squashing Commits
- Being a little lazy will bite you in the ass

![Screenshot 2021-05-13 at 15 47 52](https://user-images.githubusercontent.com/60035332/118134813-9f458600-b402-11eb-9e08-8acea254d1e2.png)

DavDan1 committed:
add semantic ui
add axios

DavDan1 committed:
add test for displaying articles,
add fixture file for fake articles

1CIM committed:
Add test for mainpage layout

1CIM committed:
Add test for display layout
setup workspace prettier
add components folder
add layout folder
add Navbar.jsx

1CIM committed:
Add style.css
WIP navbar

DavDan1 committed:
add navbar
add navigation buttons

DavDan1 committed:
Add breakingnews section

1CIM committed:
Add header with little style

1CIM committed:
Add axios baseURL

1CIM committed:
Add react-redux and redux

1CIM committed:
display articles WIP

1CIM committed:
remove redux

1CIM committed:
WIP display list of articles

DavDan1 committed:
Add footer for the application

DavDan1 committed:
add error message with status 500

DavDan1 committed:
Squash commits regarding style adding

1CIM committed:
Update test

LucasKnudsen reviewed (changes requested):
Check 'pull request' in Semaphore settings

LucasKnudsen commented on src/components/MainPage.jsx:
```diff
@@ -0,0 +1,51 @@
+  const [articles, setArticles] = useState([]);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const fetchArticles = async () => {
+    try {
+      const response = await axios.get('/articles/');
+      setArticles(response.data.articles);
+      setErrorMessage('');
+    } catch (error) {
+      if (error.response.status === 500) {
+        setErrorMessage(
+          'Servers are currently not responding, Pleas try again later'
+        );
+      } else {
+        setErrorMessage(error.message);
+      }
+    }
+  };
+  useEffect(() => {
+    fetchArticles();
+  }, []);
```
This could be moved to a module service, maybe similar to the 'controller' that we got showed earlier :)

1CIM replied:
Will take a look on that as soon as it stops being a lightning storm here...

LucasKnudsen commented on src/components/MainPage.jsx:
```diff
@@ -0,0 +1,51 @@
+          {errorMessage}
+        </Header>
+      )}
+      <BreakingNews firstArticle={firstArticle} />
```
could you put `firstArticle={articles[0]}` directly here?

1CIM replied:
yes it works! will fix it.

LucasKnudsen commented on src/components/layout/BreakingNews.jsx:
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+
+const BreakingNews = ({ firstArticle }) => {
+  if (firstArticle) {
+    return (
+      <div>
+        <div className='breaking-news' data-cy='breaking-news' placeholder>
+          <img
+          data-cy='braking-img'
+            className='braking-img'
+            src='https://images.unsplash.com/photo-1487758608033-7780b34680ac?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=3589&q=80'
+            alt=''
+          />
+          <div className='breaking-content'>
+            <h2 data-cy='breaking-title' >{firstArticle.title}</h2>
+            <h4 data-cy='breaking-teaser' >{firstArticle.teaser}</h4>
+            <p data-cy='breaking-date' >{firstArticle.date}</p>
+          </div>
+        </div>
+      </div>
+    );
+  }
+  return '';
+};
```
This mean that entire BreakingNews will not render anything if there's no articles, ye? Won't that break the design?

1CIM replied:
Somehow it goes in to the breaking news before it gives firstArticle any value so if i don't have the if() before it brakes.
Don't know other fix so please let me know if you got a better fix.

LucasKnudsen commented:
Is the footer fixed to the bottom of the screen?

1CIM committed:
Add module for getting articles
Fix Spelling mistakes in test and breakingNews file
Rename MainPage main function to MainPage
Refactor implementation of firstArticle
fix footer to display at bottom of page not window

LucasKnudsen commented on src/components/MainPage.jsx:
```diff
@@ -0,0 +1,49 @@
+  const [articles, setArticles] = useState([]);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const fetchArticles = async () => {
+    try {
+      const response = await Articles.get()
+      setArticles(response);
+      setErrorMessage('');
+    } catch (error) {
+      if (error.response.status === 500) {
+        setErrorMessage(
+          'Servers are currently not responding, Pleas try again later'
+        );
+      } else {
+        setErrorMessage(error.message);
+      }
+    }
+  };
```
If your module looks something like this, then all you need in the `useEffect` is Articles.get(functions)

```
async get(setArticles, setErrorMessage) {
  try {
    let response = await axios.get('/articles')
    setArticles(response.data.articles);
    setErrorMessage('');
  } catch (error) {
    // ...
  }
}
```

1CIM replied:
fixed

1CIM replied:
fixed

lukeperrin10 commented on src/components/MainPage.jsx:
```diff
@@ -0,0 +1,49 @@
+    try {
+      const response = await Articles.get()
+      setArticles(response);
+      setErrorMessage('');
```
why the setError on line 15?

1CIM replied:
removed

1CIM replied:
removed

1CIM committed:
Refactor Articles modeule

LucasKnudsen commented on src/components/MainPage.jsx:
```diff
@@ -0,0 +1,34 @@
+  const [errorMessage, setErrorMessage] = useState('');
+
+  useEffect(() => {
+    Articles.get(setArticles, setErrorMessage)
```
Smooth

tochman commented on .vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+  "prettier.jsxBracketSameLine": true,
+  "prettier.jsxSingleQuote": true,
+  "prettier.singleQuote": true
+}
```
Is this agreed upon by all team members? The reason I ask is that I often see different standards in your PR's

1CIM replied:
yes We just talked it over. The workspace settings should change the settings for everyone writing in the project.

tochman commented on cypress/integration/visitorCanSeeCollectionOfArticles.feature.js:
```diff
@@ -0,0 +1,20 @@
+      cy.visit('/');
+    });
+
+    it('loads fake articles', () => {
```
This is not a relevant test title. We don't care if the content comes from a fixture or not.

DavDan1 replied:
this is gone now

tochman commented on cypress/integration/visitorCanSeeMainPage.feature.js:
```diff
@@ -0,0 +1,37 @@
+      cy.get('[data-cy=breaking-img]').should('exist');
+      cy.get('[data-cy=breaking-img]').should('exist');
+      cy.get('[data-cy=breaking-title]').should('contain', 'Something');
+      cy.get('[data-cy=breaking-teaser]').should('contain', 'Something more');
```
I would update the title to include "...with a breaking content". But I also question whether this is actually in scope? I see no categories in your fixture so how do you determine that an article should be featured as "breaking"?

DavDan1 replied:
All the tests been refactored now and removed unnecessary tests and rearranged them.

tochman commented on cypress/integration/visitorCanSeeMainPage.feature.js:
```diff
@@ -0,0 +1,37 @@
+    before(() => {
+      cy.intercept('GET', 'https://fake-newzzzz.herokuapp.com/api/articles', {
+        statusCode: 500,
+        error: '500 Internal Server Error |  0     bytes\n',
```
Why `  |  0     bytes\n` ?

DavDan1 replied:
We had this same way since slow food, wasn't sure why, now it's gone :)

tochman commented on src/components/layout/BreakingNews.jsx:
```diff
@@ -0,0 +1,26 @@
+const BreakingNews = ({ firstArticle }) => {
+  if (firstArticle) {
+    return (
+      <div>
```
Why 2 div?

DavDan1 replied:
we wanted extra padding. But it works without too, so I removed it

tochman commented on src/components/layout/Footer.jsx:
```diff
@@ -0,0 +1,14 @@
+
+const Footer = () => {
+  return (
+    <div data-cy='footer' id='footer'>
```
Why the id?

DavDan1 replied:
for individual styling.

tochman commented on src/style.css:
```diff
@@ -0,0 +1,71 @@
+  margin: 50px 300px 50px;
+}
+
+#card-container {
```
This is referencing a node with an id of `card-container`, but I only see a node with a class of that name. Is this correct? Do you have either css definitions used this way? Am I missing something?

DavDan1 replied:
I've changed it to articles-container now

tochman commented on src/style.css:
```diff
@@ -0,0 +1,71 @@
+  padding: 40px;
+}
+
+.card-container{
```
Oh, I see now... Hmm, so why an id definition with the same name?

1CIM replied:
Was just a bug with our internal memory in our brain file

tochman reviewed (changes requested):
Are you in scope and the tests seem to be testing the same thing. They should be bundled into a single test file.

tochman commented:
Please fix the markdown in PR description and "What I've learned..."

DavDan1 committed:
Update test for main page layout
from exist to contain

DavDan1 committed:
remove div from articles container

DavDan1 committed:
Changes small name issue with the test
changes naming issue with css class
changes id name in first article

DavDan1 committed:
Changes small name issue with the test
changes naming issue with css class
changes id name in first article

tochman commented on src/components/layout/BreakingNews.jsx:
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+
+const BreakingNews = ({ firstArticle }) => {
```
```js
const BreakingNews = ({ firstArticle: article }) => {
```

DavDan1 replied:
Fixed  👍🏼

DavDan1 committed:
refactoring

tochman reviewed (approved):
Okay. Good to go now...
